File: python/task3c/S3_PIPE_02.py
# S3_PIPE_02 - Implement compose(*fns)

from functools import reduce

def pipe(*fns):
    return lambda x: reduce(lambda v, fn: fn(v), fns, x)

def compose(*fns):
    return lambda x: reduce(lambda v, fn: fn(v), reversed(fns), x)

# Test functions
add_one = lambda x: x + 1
double = lambda x: x * 2

# Compare pipe vs compose with same functions
print("pipe(add_one, double)(5):")
print(f"  5 -> +1 -> 6 -> *2 -> 12 = {pipe(add_one, double)(5)}")

print("\ncompose(add_one, double)(5):")
print(f"  5 -> *2 -> 10 -> +1 -> 11 = {compose(add_one, double)(5)}")

# compose applies right-to-left, pipe applies left-to-right

File: js/task3c/S3_PIPE_02.js
// S3_PIPE_02 - compose()
// right-to-left function composition

const compose = (...fns) => x => {
  let result = x;
  for (let i = fns.length - 1; i >= 0; i--) {
    result = fns[i](result);
  }
  return result;
};

// tests
const addOne = x => x + 1;
const double = x => x * 2;
const square = x => x * x;

const composed = compose(square, double, addOne);

console.log("compose(square, double, addOne)(3):");
console.log("  Executes right-to-left: addOne -> double -> square");
console.log("  3 -> +1 -> 4 -> *2 -> 8 -> ^2 -> 64");
console.log("  Result:", composed(3)); // 64

// compare with pipe - same result if order is reversed
console.log("\nSame as pipe(addOne, double, square)(3)");
